### ISSUE TITLE
Unreachable call of non-const fn from const fn is not a compilation error

### ISSUE THREAD

**Issue by tonowak** [rust-lang/rust#110831]
I tried this code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=37a476cf426f9705dac5a0ea0f3e42e2)):

```rust
fn non_const_fn(value: i64) {}

const fn witness() {
    non_const_fn(todo!());
}
```

I expected to see this happen: a compilation error. Calling a non-const function (even though it's unreachable) should produce a compilation error, just like it would with a `i64` value instead of `todo!()`.

The compilation error with a value instead of `todo!()`:
```
error[E0015]: cannot call non-const fn `non_const_fn` in constant functions
 --> src/lib.rs:4:5
  |
4 |     non_const_fn(0);
  |     ^^^^^^^^^^^^^^^
  |
```

Instead, this happened: the code compiles (with a warning `unreachable call`).

### Meta

`rustc --version --verbose`:
```
rustc 1.67.0 (fc594f156 2023-01-24)
binary: rustc
commit-hash: fc594f15669680fa70d255faec3ca3fb507c3405
commit-date: 2023-01-24
host: x86_64-unknown-linux-gnu
release: 1.67.0
LLVM version: 15.0.6

```

```
rustc 1.71.0-nightly (f5559e338 2023-04-24)
binary: rustc
commit-hash: f5559e338256f17ada6d82b429acc2dbd8facc9c
commit-date: 2023-04-24
host: x86_64-unknown-linux-gnu
release: 1.71.0-nightly
LLVM version: 16.0.2
```

<details><summary>Backtrace</summary>
<p>

</p>
</details>
